okhttp/src/main/java/okhttp3/Request.kt

```diff
@@ -0,0 +1,236 @@
+     *
+     * @throws IllegalArgumentException if [url] is not a valid HTTP or HTTPS URL. Avoid this
+     * exception by calling [HttpUrl.parse]; it returns `null` for invalid URLs.
+     */
+    open fun url(url: String): Builder {
+      // Silently replace web socket URLs with HTTP URLs.
+      val finalUrl: String = when {
+        url.regionMatches(0, "ws:", 0, 3, ignoreCase = true) -> {
+          "http:${url.substring(3)}"
+        }
+        url.regionMatches(0, "wss:", 0, 4, ignoreCase = true) -> {
+          "https:${url.substring(4)}"
+        }
+        else -> url
+      }
+
+      return url(HttpUrl[finalUrl])
+    }
+
+    /**
+     * Sets the URL target of this request.
+     *
+     * @throws IllegalArgumentException if the scheme of [url] is not `http` or `https`.
+     */
+    // Code review TODO - this never checked the scheme before like the doc says. Should it?
+    open fun url(url: URL) = url(HttpUrl[url.toString()])
+
+    /**
+     * Sets the header named [name] to [value]. If this request already has any headers
+     * with that name, they are all replaced.
+     */
+    open fun header(name: String, value: String) = apply {
+      headers[name] = value
+    }
+
+    /**
+     * Adds a header with [name] and [value]. Prefer this method for multiply-valued
+     * headers like "Cookie".
+     *
+     * Note that for some headers including `Content-Length` and `Content-Encoding`,
+     * OkHttp may replace [value] with a header derived from the request body.
+     */
+    open fun addHeader(name: String, value: String) = apply {
+      headers.add(name, value)
+    }
+
+    /** Removes all headers named [name] on this builder. */
+    open fun removeHeader(name: String) = apply {
+      headers.removeAll(name)
+    }
+
+    /** Removes all headers on this builder and adds [headers]. */
+    open fun headers(headers: Headers) = apply {
+      this.headers = headers.newBuilder()
+    }
+
+    /**
+     * Sets this request's `Cache-Control` header, replacing any cache control headers already
+     * present. If [cacheControl] doesn't define any directives, this clears this request's
+     * cache-control headers.
+     */
+    open fun cacheControl(cacheControl: CacheControl): Builder {
+      val value = cacheControl.toString()
+      return when {
+        value.isEmpty() -> removeHeader("Cache-Control")
+        else -> header("Cache-Control", value)
+      }
+    }
+
+    open fun get() = method("GET", null)
+
+    open fun head() = method("HEAD", null)
+
+    open fun post(body: RequestBody) = method("POST", body)
+
+    @JvmOverloads
+    open fun delete(body: RequestBody? = Util.EMPTY_REQUEST) = method("DELETE", body)
+
+    open fun put(body: RequestBody) = method("PUT", body)
+
+    open fun patch(body: RequestBody) = method("PATCH", body)
+
+    open fun method(method: String, body: RequestBody?): Builder = apply {
+      require(method.isNotEmpty()) {
+        "method.isEmpty() == true"
```

Comment:
weird that the exception message is changing here. The `== true` part weirds me out also

Reply:
I was matching the referenced check, would you rather keep the `.length == 0`?